treeherder/auth/backends.py

```diff
@@ -51,32 +65,54 @@ def _extract_email_from_clientid(self, client_id):
             "No email found in clientId: '{}'".format(client_id))
 
     def authenticate(self, request):
-
-        auth_header = request.META.get("HTTP_TCAUTH", None)
-        host = request.get_host().split(":")[0]
-        port = int(request.get_port())
-
-        if not auth_header:
-            # Doesn't have the right params for this backend.  So just
-            # skip and let another backend have a try at it.
-            return None
-
-        tc_auth = Auth()
-        # https://docs.taskcluster.net/reference/platform/taskcluster-auth/references/api#authenticateHawk
-        # https://github.com/taskcluster/taskcluster-client.py#authenticate-hawk-request
-        result = tc_auth.authenticateHawk({
-            "authorization": auth_header,
-            "host": host,
-            "port": port,
-            "resource": reverse("auth-login"),
-            "method": "get",
-        })
-
-        if result["status"] != "auth-success":
-            logger.warning("Error logging in: {}".format(result["message"]))
-            raise TaskclusterAuthException(result["message"])
-
-        client_id = result["clientId"]
+        token = self._get_token_auth_header(request)
+        expires_at_in_milliseconds = self._get_session_expiry(request)
+        now_in_milliseconds = int(round(time.time() * 1000))
+
+        # The Django user session expiration should be set to match the expiry time of the Auth0 token
+        request.session.set_expiry((expires_at_in_milliseconds - now_in_milliseconds) / 1000)
+
+        # JWT Validator
+        # Per https://auth0.com/docs/quickstart/backend/python/01-authorization#create-the-jwt-validation-decorator
+        jsonurl = urlopen('https://' + AUTH0_DOMAIN + '/.well-known/jwks.json')
+        jwks = json.loads(jsonurl.read())
+        unverified_header = jwt.get_unverified_header(token)
+
+        rsa_key = {}
+        for key in jwks["keys"]:
+            if key["kid"] == unverified_header["kid"]:
+                rsa_key = {
+                    "kty": key["kty"],
+                    "kid": key["kid"],
+                    "use": key["use"],
+                    "n": key["n"],
+                    "e": key["e"]
+                }
```

Comment:
OK, so @andrewkrug pointed me to https://github.com/mozilla-iam/sso-dashboard/blob/master/dashboard/auth.py which uses josepy, but python-jose should be similar.  I worry a lot about the `urlopen` call in this PR especially -- what if that 503's or we get rate-limited?  He let me know that the public key at `/.well-knon/jwks.json` is under our (Mozilla's) control and that changing it would be a big thing with lots of notice in advance.  So, I think we could configure a single JWK statically in the Treeherder config, and verify against that, similar to what is happening in the dashboard code linked above.

Reply:
OK, this is much better now :)